proposals/0028/README.md

```diff
@@ -0,0 +1,121 @@
+
+## Goals and not Goals
+
+Goals:
+
+* Enable network booting in environments using trunked ports without a native vlan.
+
+Non Goals:
+
+* Custom workflow actions for configuring OS level VLAN tagging.
+* Enabling of VLAN tagging via a machine's BMC with Rufio.
+
+## Content
+
+### Components
+
+The following is an initial estimate of the changes that will be needed in order to enable the VLAN tagging functionality.
+
+| Code Repo | Required Changes |
+| --------- | ---------------- |
+| ipxedust  | add `vcreate` capability to embedded iPXE script |
+| boots     | 1. add VLAN ID to DHCP opt 43.116 2. add `vlanid=` to kernel command line in the auto.ipxe script |
+| dhcp      | add VLAN ID to DHCP opt 43.116 |
+| tink      | add VLAN ID to hardware spec for interfaces |
+| hook      | add virtual interface with VLAN tagging capability based on `vlanid=` found in `/proc/cmdline` |
+
+### User Experience
+
+In order to enable VLAN tagging support from the Tinkerbell stack side, a user would only need to add the VLAN ID to the hardware spec for the interface. Generally, there will still be additional steps a user would need to take. Some outside of the Tinkerbell stack.
+
+1. Tag interface(s) on switches for which the physical machine is connected.
+2. Enable and set a VLAN ID in the out-of-band software for the physical machine.
+3. Create a template workflow action to add VLAN tagging to an interface for the OS being installed.
+
+All other processes around workflows, etc would be the same as they are currently.
+
+### Failure modes
+
+Failures to read the VLAN ID from DHCP option 43.116 in iPXE will not block iPXE from booting. The VLAN ID will just not be set and not propagated through to Hook.
+
+### Tradeoffs
+
+This feature adds complexity to the stack. Testing will most likely be manual and very dependent on the environment.
+
+## System-context-diagram
+
+![System-context-diagram](vlan_tag_feature.png)
+
+## APIs
+
+Example Hardware spec with added vlan id.
+
+```yaml
+apiVersion: "tinkerbell.org/v1alpha1"
+kind: Hardware
+metadata:
+  name: sm01
+  namespace: default
+spec:
+  disks:
+    - device: /dev/nvme0n1
+  metadata:
+    facility:
+      facility_code: onprem
+    manufacturer:
+      slug: supermicro
+    instance:
+      userdata: ""
+      hostname: "sm01"
+      id: "3c:ec:ef:4c:4f:54"
+      operating_system:
+        distro: "ubuntu"
+        os_slug: "ubuntu_20_04"
+        version: "20.04"
+  interfaces:
+    - dhcp:
+        arch: x86_64
+        hostname: sm01
+        ip:
+          address: 172.16.10.100
+          gateway: 172.16.10.1
+          netmask: 255.255.255.0
+        lease_time: 86400
+        mac: 3c:ec:ef:4c:4f:54
+        name_servers:
+          - 172.16.10.1
+          - 10.1.1.11
+        uefi: true
+        vlan_id: "12"
+      netboot:
+        allowPXE: true
+        allowWorkflow: true
+```
+
+## Alternatives
+
+Don't support VLAN tagging.
```

Comment:
One of the items discussed in the community calls was wether or not there was an existing standard for DHCP declaring VLAN configuration. @jacobweinstock pointed out that there were none and that the nearest results were vendor specific and tailored to VOIP solutions (some using option 43, other 132):

https://www.iana.org/assignments/bootp-dhcp-parameters/bootp-dhcp-parameters.xhtml
Option 132 - IEEE 802.1Q VLAN ID

Here are some examples:
* https://www.sonicwall.com/support/knowledge-base/how-to-configure-dhcp-option-132-to-get-dhcp-lease-from-vlan/220309054946100/
* https://wiki.unify.com/wiki/VLAN_ID_Discovery_over_DHCP#Example_configuration_of_an_Cisco_Switch
* https://www.teradici.com/web-help/TER1206003/4.x-4.9/07_HowTo/Config_VLAN.htm
* https://wiki.freepbx.org/display/PHON/DHCP+VLAN+Option+132

Reply:
Hey @displague , thanks for this detail! I've updated the proposal to include this info.